File: api/golang/core/lib/starlark_run_config/starlark_run_config.go
package starlark_run_config

import (
	"github.com/kurtosis-tech/kurtosis/api/golang/core/kurtosis_core_rpc_api_bindings"
)

const (
	defaultRelativePathToMainFile = ""
	defaultMainFunctionName       = ""
	defaultSerializedParams       = "{}"
	defaultDryRun                 = false
	defaultParallelism            = 4
	defaultCloudInstanceId        = ""
	defaultCloudUserId            = ""
	defaultImageDownload          = kurtosis_core_rpc_api_bindings.ImageDownloadMode_missing
	defaultNonBlockingMode        = false
	defaultGitHubAuthToken        = ""
	defaultParallel               = false
)

var defaultExperimentalFeatureFlags = []kurtosis_core_rpc_api_bindings.KurtosisFeatureFlag(nil)

type StarlarkRunConfig struct {
	RelativePathToMainFile   string
	MainFunctionName         string
	SerializedParams         string
	DryRun                   bool
	Parallelism              int32
	ExperimentalFeatureFlags []kurtosis_core_rpc_api_bindings.KurtosisFeatureFlag
	CloudInstanceId          string
	CloudUserId              string
	ImageDownload            kurtosis_core_rpc_api_bindings.ImageDownloadMode
	NonBlockingMode          bool
	GitHubAuthToken          string
	Parallel                 bool
}

type starlarkRunConfigOption func(*StarlarkRunConfig)

func NewRunStarlarkConfig(opts ...starlarkRunConfigOption) *StarlarkRunConfig {
	config := &StarlarkRunConfig{
		RelativePathToMainFile:   defaultRelativePathToMainFile,
		MainFunctionName:         defaultMainFunctionName,
		SerializedParams:         defaultSerializedParams,
		DryRun:                   defaultDryRun,
		Parallelism:              defaultParallelism,
		ExperimentalFeatureFlags: defaultExperimentalFeatureFlags,
		CloudInstanceId:          defaultCloudInstanceId,
		CloudUserId:              defaultCloudUserId,
		ImageDownload:            defaultImageDownload,
		NonBlockingMode:          defaultNonBlockingMode,
		GitHubAuthToken:          defaultGitHubAuthToken,
		Parallel:                 defaultParallel,
	}

	for _, opt := range opts {
		opt(config)
	}

	return config
}

func WithRelativePathToMainFile(relativePathToMainFile string) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.RelativePathToMainFile = relativePathToMainFile
	}
}

func WithMainFunctionName(mainFunctionName string) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.MainFunctionName = mainFunctionName
	}
}

func WithSerializedParams(serializedParams string) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.SerializedParams = serializedParams
	}
}

func WithDryRun(dryRun bool) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.DryRun = dryRun
	}
}

func WithParallelism(parallelism int32) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.Parallelism = parallelism
	}
}

func WithExperimentalFeatureFlags(experimentalFeatureFlags []kurtosis_core_rpc_api_bindings.KurtosisFeatureFlag) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.ExperimentalFeatureFlags = experimentalFeatureFlags
	}
}

// Deprecated: This isn't used anymore and the value passed will not be read anywhere
func WithCloudInstanceId(cloudInstanceId string) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.CloudInstanceId = cloudInstanceId
	}
}

// Deprecated: This isn't used anymore and the value passed will not be read anywhere
func WithCloudUserId(cloudUserId string) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.CloudUserId = cloudUserId
	}
}

func WithImageDownloadMode(imageDownloadMode kurtosis_core_rpc_api_bindings.ImageDownloadMode) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.ImageDownload = imageDownloadMode
	}
}

func WithNonBlockingMode(nonBlockingMode bool) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.NonBlockingMode = nonBlockingMode
	}
}

func WithGitHubAuthToken(githubAuthToken string) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.GitHubAuthToken = githubAuthToken
	}
}

func WithParallel(parallel bool) starlarkRunConfigOption {
	return func(config *StarlarkRunConfig) {
		config.Parallel = parallel
	}
}

File: api/typescript/src/core/lib/enclaves/starlark_run_config.ts
import {KurtosisFeatureFlag} from "../../kurtosis_core_rpc_api_bindings/api_container_service_pb";

type StarlarkRunConfigOption = (starlarkRunConfig: StarlarkRunConfig) => void;

const DEFAULT_RELATIVE_PATH_TO_MAIN_FILE = ""
const DEFAULT_MAIN_FUNCTION_NAME = ""
const DEFAULT_SERIALIZED_PARAMS = "{}"
const DEFAULT_DRY_RUN = false
const DEFAULT_PARALLELISM = 4
const DEFAULT_EXPERIMENTAL_FEATURE_FLAGS = Array<KurtosisFeatureFlag>()
const DEFAULT_CLOUD_INSTANCE_ID = ""
const DEFAULT_CLOUD_USER_ID = ""

export class StarlarkRunConfig {
    public relativePathToMainFile: string
    public mainFunctionName: string
    public serializedParams: string
    public dryRun: boolean
    public parallelism: number
    public experimentalFeatureFlags: Array<KurtosisFeatureFlag>
    public cloudInstanceId: string
    public cloudUserId: string

    constructor(...options: StarlarkRunConfigOption[]) {
        this.relativePathToMainFile = DEFAULT_RELATIVE_PATH_TO_MAIN_FILE
        this.mainFunctionName = DEFAULT_MAIN_FUNCTION_NAME
        this.serializedParams = DEFAULT_SERIALIZED_PARAMS
        this.dryRun = DEFAULT_DRY_RUN
        this.parallelism = DEFAULT_PARALLELISM
        this.experimentalFeatureFlags = DEFAULT_EXPERIMENTAL_FEATURE_FLAGS
        this.cloudInstanceId = DEFAULT_CLOUD_INSTANCE_ID
        this.cloudUserId = DEFAULT_CLOUD_USER_ID

        for (const option of options) {
            option(this)
        }
    }

    public static WithRelativePathToMainFile(relativePathToMainFile: string): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.relativePathToMainFile = relativePathToMainFile
        }
    }

    public static WithMainFunctionName(mainFunctionName: string): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.mainFunctionName = mainFunctionName
        }
    }

    public static WithSerializedParams(serializedParams: string): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.serializedParams = serializedParams
        }
    }

    public static WithDryRun(dryRun: boolean): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.dryRun = dryRun
        }
    }

    public static WithParallelism(parallelism: number): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.parallelism = parallelism
        }
    }

    public static WithExperimentalFeatureFlags(experimentalFeatureFlags: Array<KurtosisFeatureFlag>): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.experimentalFeatureFlags = experimentalFeatureFlags
        }
    }

    public static WithCloudInstanceId(cloudInstanceId: string): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.cloudInstanceId = cloudInstanceId
        }
    }

    public static WithCloudUserID(cloudUserId: string): StarlarkRunConfigOption {
        return (config: StarlarkRunConfig): void => {
            config.cloudUserId = cloudUserId
        }
    }
}